docker: reduce layers, remove unused packages

Strubbl opened the pull request:
this is just a draft, open for discussion

Strubbl committed:
docker: reduce layers

Strubbl commented:
image size is reduced from 1.42 GB to 1.01 GB

dansup commented:
cc @shleeable

shleeable commented:
Hey @Strubbl, those unused packages are definitely used... (I mean the postgres one isn't by default but we will get Postgres as the default DB one day)

The one big RUN thing is fine I guess after that dockerfile has been tested... but being able to run using cached images to speed up the build was my goal during testing. that said, it only takes a few minutes to build at most.

so Cool?

shleeable commented:
I'll recommend a few changes in a bit

Strubbl commented:
I am successfully running this setup for a month now

dm commented on contrib/docker/Dockerfile.apache:
```diff
@@ -4,21 +4,21 @@ FROM php:7.4-apache-buster
+ENV COMPOSER_VERSION 1.10.6
 ENV COMPOSER_HOME /var/www/.composer
+ENV PATH="~/.composer/vendor/bin:./vendor/bin:${PATH}"
+COPY . /var/www/
```
Best practice: I'd still leave the `COPY` command at the end of the whole image setup `RUN` command. Followed by the 
So code changes in the filesystem don't trigger the whole image rebuild.

Followed by another single RUN command of the remaining composer/code build commands

Strubbl replied:
thanks, i updated the commit accordingly.

Strubbl commented on contrib/docker/Dockerfile.apache:
```diff
@@ -33,74 +32,65 @@ RUN apt-get install -y --no-install-recommends \
-RUN composer global remove hirak/prestissimo
-RUN rm -rf html && ln -s public html
-
+RUN cp -r storage storage.skel \
```
why is the actual storage copied to the skeleton dir. is that a bug?

dm replied:
Yeah, I find this weird as well...
I'm quite confused, likely just going to recreate a proper Docker build for my fork

hnrd replied:
I guess it's to pull in future updates to the shipped storage to your local storage volume. 
`storage` here is the one from the git checkout, where `storage` on your running container is your volume. 
By copying `storage.skel` back to `storage` in the entrypoint, it automatically updates the local storage in case the image ships with a new subdirectory or something.

Strubbl replied:
but `storage` is the source and its content is copied to the target `storage.skel`

shleeable replied:
Nobody knows why this exists but everybody is too scared to change it ;)

leave it as is if we want to get this merged and we can test later.

hnrd replied:
Yes, it's the source from github. That's the part we want to keep to update our volumes from past version.

Let me try to construct an example:
- We ship `storage/` populated with some subdirectories: https://github.com/pixelfed/pixelfed/tree/dev/storage
- Next step, you download the repo, build your image, spin up your container and mount storage as a volume, because you want to keep your uploads.
- On Container+Volume creation, the content of `storage/ (github)` get copied to `storage/ (your volume)`, but only on the first time
- Now, on a future update, `storage/ (github)` introduces a new subdirectory
- When you update your existing deployment, and mount your old `storage/ (volume)` back, you would be missing that new subdirectory

That's why we always copy `storage/ (github)` to that skel directory on image creation, and back on container creation (in the entrypoint). So you'll always get all of github-storages content in your volume.

If you want we can continue this on IRC, I'm hansenerd, that'd be likely a lot faster ;)

hnrd replied:
A real world example that I just noticed: If cities.json would be updated in this repo, it would never reach your existing volume if it weren't for the skel workaround.

dansup commented:
Is this ready to merge? cc @Strubbl @shleeable

Strubbl commented:
i would like to clarify the line "RUN cp -r storage storage.skel" before merging

shleeable commented:
Nobody knows why this exists but everybody is too scared to change it ;)

leave it as is if you want to get this merged and we can test later.

Strubbl commented:
I just updated the composer and made a short comment with a link to this `storage` discussion

shleeable commented on contrib/docker/Dockerfile.apache:
```diff
@@ -33,74 +32,66 @@ RUN apt-get install -y --no-install-recommends \
-RUN docker-php-ext-install pdo_mysql pdo_pgsql pdo_sqlite
-
+  && docker-php-ext-install pdo_mysql \
+#pdo_pgsql pdo_sqlite \
```
What's the deal with commenting the pgsql? We'll likely need that.

Strubbl replied:
also sqlite? The docker-compose uses mysql. That's the reason. If we need it in future we can uncomment this, don't we?

hnrd replied:
I second including the drivers needed to run the stack as defined in this repo (mysql, that is) and not including (at this time) unneeded drivers.

dansup commented:
Is this ready for me to test?

Strubbl commented:
yes, i am using this branch already with my instance

Strubbl commented:
the dockerfile needs an update w.r.t. #2220 i think. i am currently working on this

Strubbl commented:
just updated the dockerfile and added the fixes recommended in #2220. 

the following lines in the file are new (not in that direct order):
```
ENV COMPOSER_MEMORY_LIMIT -1
ARG DEBIAN_FRONTEND=noninteractive
  && apt-get upgrade -y \
  && composer update --no-dev --prefer-dist --no-interaction --no-ansi \
```

Strubbl commented:
i reduced the number of layers by 3, because i summarized the 4 ENV lines to only one

dansup commented:
@Strubbl Is this ready to merge?

shleeable commented:
Fine with me... I'll boot another demo instance and we can debug after the next release.

Strubbl commented:
> @Strubbl Is this ready to merge?

yes, please. :)

Strubbl commented:
updated to composer 1.10.9

Strubbl commented:
added mariadb-client, so that mysql-dump can be utilized in the app container

Strubbl commented:
rebased against staging. ready to merge.
